File: leetcode-cn/0494.0_Target_Sum.py
'''
59 / 138 个通过测试用例
状态：超出时间限制
提交时间：几秒前
最后执行的输入：
[9,28,50,9,34,48,2,50,38,10,5,16,44,5,48,21,38,17,21,49]
20
'''

class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        N = len(nums)
        self.ways = 0

        #@lru_cache(None) # will be wrong
        def dfs(i, N, cur, target):
            if i == N:
                if cur == target:
                    self.ways += 1
                return

            # add
            dfs(i + 1, N, cur + nums[i], target)
            # minus
            dfs(i + 1, N, cur - nums[i], target)
        
        dfs(0, N, 0, target)
        return self.ways


'''
DFS + @cache
Time: O(2^N)
Space: O(N)

执行用时：304 ms, 在所有 Python3 提交中击败了40.25% 的用户
内存消耗：46.1 MB, 在所有 Python3 提交中击败了5.04% 的用户
'''

class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        N = len(nums)
        
        @cache
        def dfs(i, N, cur, target):
            if i == N:
                if cur == target:
                    return 1
                else:
                    return 0

            # add
            add = dfs(i + 1, N, cur + nums[i], target)
            # minus
            minus = dfs(i + 1, N, cur - nums[i], target)
            return add + minus
        
        return dfs(0, N, 0, target)

File: java/0494.0_Target_Sum.java
/*
DFS

执行用时：658 ms, 在所有 Java 提交中击败了17.82% 的用户
内存消耗：35.8 MB, 在所有 Java 提交中击败了88.07% 的用户
*/

class Solution {
    int ans = 0;
    public int findTargetSumWays(int[] nums, int target) {
        dfs(nums, target, 0, 0);
        return ans;
    }
    void dfs(int[] nums, int t, int u, int cur) {
        if (u == nums.length) {
            ans += cur == t ? 1 : 0;
            return;
        }
        dfs(nums, t, u + 1, cur + nums[u]);
        dfs(nums, t, u + 1, cur - nums[u]);
    }
}



/*
DFS

执行用时：696 ms, 在所有 Java 提交中击败了12.28% 的用户
内存消耗：35.6 MB, 在所有 Java 提交中击败了98.55% 的用户
*/

class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        return dfs(nums, target, 0, 0);
    }
    int dfs(int[] nums, int t, int u, int cur) {
        if (u == nums.length) {
            return cur == t ? 1 : 0;
        }
        int add = dfs(nums, t, u + 1, cur + nums[u]);
        int minus = dfs(nums, t, u + 1, cur - nums[u]);
        return add + minus;
    }
}
